[processor/schematransformerprocessor] Initial PR

MovieStoreGuy opened the pull request:
**Description:**
As part of the [accepted specification](https://github.com/open-telemetry/opentelemetry-specification/blob/e16e1a781bdf62a8fcfa1a12963844749de23e7a/specification/schemas/overview.md#collector-assisted-schema-transformation), there needs to be a processor within the collector that is able to allow for transformations of semantic convention versions . This is my initial draft, I wrote this before https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8344/files shared their solution to it.

**Link to tracking Issue:**
open-telemetry/opentelemetry-collector-contrib#8495
Original raised in the collector but should be shifted to the contrib a I can tell.

**Testing:**
It currently does nothing and there is the initial testing set up within the components.

**Documentation:** <Describe the documentation added.>
Added a minimal README, will expand upon it once more work has been added to the collector.

MovieStoreGuy commented:
I have avoided adding any implementation detail here to keep the high level PR in, this PR can _should_ be fine to merge in and run (since it will only just validate the configuration, passthrough exactly what it receives).

Not entirely sure how to resolve the issue within the `gomoddownload` step, @codeboten if you have any ideas please let me know :) 

After this initial PR goes in, the next logical flow I have in mind is:
- Figure out the best way to cache schemas
- Determine schema versions and applying changes with minimal efforts
- Adopt @tigrannajaryan compiled schemas to reduce runtime efforts
- Add in directory extension support to write schemas to disk

Aneurysm9 commented on go.mod:
```diff
@@ -685,6 +686,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/processor/tail
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor => ./processor/transformprocessor
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/processor/schematransformerprocessor => ./processor/schematransformerprocessor/
```
```suggestion
replace github.com/open-telemetry/opentelemetry-collector-contrib/processor/schematransformerprocessor => ./processor/schematransformerprocessor
```

This might fix your mod download issue.

MovieStoreGuy replied:
Oh are you serious? Let me a have a look, super silly if that is the cause.

MovieStoreGuy commented on processor/schematransformerprocessor/transformer_test.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+func TestTrasnformerProcessing(t *testing.T) {
+	t.Parallel()
```
I should probably finish this test...

Aneurysm9 commented:
You'll also need to add a `replace` directive to [`cmd/configschema/go.mod`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7a4fac483f475b8eed8b1e28c3970b6e90699582/cmd/configschema/go.mod) as it depends on the root module and `replace` directives are only honored from the current `go.mod` for any operation.  It's super annoying and we're working on a tool to make this a non-issue, but until then some hand-crafting is necessary.

MovieStoreGuy commented:
Thanks @Aneurysm9, it looks like it managed to get past the `gomoddownload` part and getting onto the meaningful tests :)

MovieStoreGuy commented:
Nobody move... the load tests somehow passed.

@codeboten , @jpkrohling, @Aneurysm9  , if you have time to have a look over this pull request?

jpkrohling commented on processor/schematransformerprocessor/testdata/schema.yml:
```diff
@@ -0,0 +1,110 @@
+      # Definitions that apply to all data types.
+      changes:
+        # Transformations to apply when converting from version 1.0.0 to 1.1.0.
+        - rename_attributes:
```
That looks nice, where is this coming from? Would we be maintaining this here in this repo?

MovieStoreGuy replied:
This was the example taken from the specification https://github.com/open-telemetry/opentelemetry-specification/blob/e16e1a781bdf62a8fcfa1a12963844749de23e7a/specification/schemas/overview.md#collector-assisted-schema-transformation 

The idea here is to use this within future testing one the schema transformer is ready.

MovieStoreGuy replied:
As for who owns this schema translation files, I believe that is Instrumentation group led by @tedsuo

Aneurysm9 replied:
The schemas themselves live in https://github.com/open-telemetry/opentelemetry-specification/tree/main/schemas

tigrannajaryan replied:
Schema files are part of the Otel specification and are published at `https://opentelemetry.io/schemas/*` for every new specification release. The specification SIG owns these files.

jpkrohling reviewed (commented):
Sorry about that, but we just had a new release, so, the go.mod/sum files here have to be changed.

MovieStoreGuy commented on processor/schematransformerprocessor/internal/schema/indentifer.go:
```diff
@@ -0,0 +1,68 @@
+	return NewIdentifier(ident)
+}
+
+func NewIdentifier(s string) (*Identifier, error) {
```
Yes, that is an easy to do.

MovieStoreGuy commented on processor/schematransformerprocessor/factory_test.go:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schematransformerprocessor_test
```
Yes, since the transformer is currently just an empty struct, I wasn't exactly sure how the transformer would look like so I didn't want to burden reviewers with more work than required.

MovieStoreGuy commented on processor/schematransformerprocessor/config.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+var (
+	ErrInvalidFileFormatVersion = errors.New("invalid file format version")
```
They don't need to be, this was more a force of habit of mine to make it easier to check / mock errors from the package.

MovieStoreGuy commented on processor/schematransformerprocessor/internal/schema/indentifer_test.go:
```diff
@@ -0,0 +1,66 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema_test
```
Sorry, do you mind giving an example?

jpkrohling replied:
I might certainly be wrong, but I just picked a random component and it's indeed using the same package for both the tests and the test subject:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b6a9b30160f6e82d61b95f11568be8f415a8c93d/processor/cumulativetodeltaprocessor/factory_test.go#L15

Aneurysm9 replied:
Yes, we typically test from the same package to ensure that unexported types and functions can be tested in isolation where needed.  Occasionally it may be useful to test from an external package perspective, but that should generally be reserved for cases such as validating that something is/isn't possible from outside the package under test.

MovieStoreGuy replied:
Right! Sorry, Brain failed to engage :) 

Yeah, that is completely fine. I typically prefer doing it this way so that changes to functions or signatures are easily caught and super obvious. 

Happy to move the package to use the exported package instead of the test package to make sure all code is covered.

MovieStoreGuy commented:
Lame, the loadtest flaked out on me.

MovieStoreGuy commented:
The stairs have aligned once again! Once I get a chance, I will have a stab to see if I can help reduce the flake in the those tests.

jpkrohling commented on cmd/configschema/go.mod:
```diff
@@ -303,6 +303,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourcedetectionprocessor v0.47.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourceprocessor v0.47.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor v0.47.0 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/schematransformerprocessor v0.0.0-00010101000000-000000000000 // indirect
```
I'm not sure about this change. Would it get updated to the same version as the others when they are bumped?

Aneurysm9 replied:
Yes, that'll get updated with the next release prep PR after it is merged.

jpkrohling commented on processor/schematransformerprocessor/internal/schema/indentifer.go:
```diff
@@ -0,0 +1,83 @@
+// Copyright  The OpenTelemetry Authors
```
Sorry for noticing this only now, but the file name should be `identifier.go`. It currently has an extra "n" after the initial "i".

jpkrohling commented on processor/schematransformerprocessor/internal/schema/indentifer_test.go:
```diff
@@ -0,0 +1,89 @@
+// Copyright  The OpenTelemetry Authors
```
This file name should also be fixed.

tigrannajaryan commented on processor/schematransformerprocessor/config.go:
```diff
@@ -0,0 +1,84 @@
+type Config struct {
+	config.ProcessorSettings `mapstructure:",squash"`
+
+	// SchemaURL defines where to download the published schema file
```
The purpose of this field is unclear. Why does the user need to specify this? What will the processor do with this schema file?

MovieStoreGuy replied:
So my intentions for this field to be were to obtain the schema translation file of version `x`, then allow for semantic conversions from 1.0.0 to `x` (assuming that published version x is the most latest version to perform translations to, I will need to reread the spec to make sure).

tigrannajaryan commented on processor/schematransformerprocessor/config.go:
```diff
@@ -0,0 +1,84 @@
+	// SchemaFileVersion defines what the expected file format is
+	// and assist in making loading easier, this value isn't expected to change
+	// from the default is added know to make it aware that file versions matter.
+	SchemaFileVersion string `mapstructure:"schema_file_version"`
```
This seems unnecessary. The file format version is already a field in the schema file. How does this setting here help with loading?

MovieStoreGuy replied:
If this field were to be used, it would remove the need to potential pass the bytes more than once since yaml formats don't enforce order of the dictionary.

For example, if versions was declared inside the file version then a rewind has to be done or reprocessing that map of type `interface`. 

This is where I was going with this however, it does seem like a micro optimisation since these files are only intended to be parsed once and then compiled.

tigrannajaryan commented on processor/schematransformerprocessor/config.go:
```diff
@@ -0,0 +1,84 @@
+	SchemaFileVersion string `mapstructure:"schema_file_version"`
+
+	// SchemaTarget defines what the outputted semantic convention version
+	// should be. The Schema Identifier must be within the versions defined with the schema
```
>The Schema Identifier must be within the versions defined with the schema

This part is not clear.

MovieStoreGuy replied:
Right, so part of the spec it says that the `target schema_url` is defined so that all processed telemetry matches that output but this value also needs to be defined in the schema translation file obtained from `SchemaURL`.

tigrannajaryan commented on processor/schematransformerprocessor/README.md:
```diff
@@ -0,0 +1,25 @@
+
+Supported Pipelines: traces, metrics, logs
+
+The _Schema Transformer Processor_ is used to convert existing telemetry data or signals to the most recent targeted version of the semantic convention defined as part of the configuration. The latest schema translation file will be bundled as part of the collector build to avoiding having to fetch it at runtime. 
```
>to the most recent targeted version

I think we need to allow the user to choose to which version to transform to, not just to the most recent.

Aneurysm9 replied:
Translation can go forward and backward and the user should definitely be able to specify the target version.  Indeed, I expect a common use case will be to ensure that telemetry coming in with a new schema version is adapted to the currently expected, but older, schema version.

MovieStoreGuy replied:
Yes, I did forget to go back and update this part after reading through the spec in more detail. Let me fix this up in a follow up commit :)

tigrannajaryan commented on processor/schematransformerprocessor/README.md:
```diff
@@ -0,0 +1,25 @@
+processors:
+    schema_transformer:
+        schema:
+            url: https://opentelemetry.io/schemas/1.9.0
+            target: https://opentelemtry.io/schemas/1.5.0
```
The settings names don't match what is `Config` struct. Also, the purpose of `url` is not clear.

MovieStoreGuy replied:
This is me being terrible at documentation, let me go back and chase this up :)

tigrannajaryan commented on processor/schematransformerprocessor/testdata/config.yml:
```diff
@@ -0,0 +1,29 @@
+processors:
+  schema_transformer:
+    # schema_url defines the remote location of the published
+    # schema translation to convert from an older semantic convention
```
>to convert from an older semantic convention

I think we need to support converting both ways: from an older to newer and from newer to older.

MovieStoreGuy replied:
This also reads as an incomplete thought that didn't make through to the keyboard.
I'll also address this as well.

tigrannajaryan commented on processor/schematransformerprocessor/README.md:
````diff
@@ -0,0 +1,25 @@
+
+```yaml
+processors:
+    schema_transformer:
````
Can we simplify the name to `schema`? Is there anything else that we can do with the schema other than transform it?

MovieStoreGuy replied:
For this, I was following what was written into the spec however, the spec itself uses two seperate names for the component so I guess it shouldn't matter too much.

MovieStoreGuy commented:
Sorry @tigrannajaryan,

Finally got a chance to get back to this :)
If you have a moment to look over this it would be greatly appreciated 🙏🏽

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,76 @@
+type Config struct {
+	config.ProcessorSettings `mapstructure:",squash"`
+
+	// SchemaURL defines a url where the schema can be loaded from
+	// and then used to then evaluate translations between the configured version
+	SchemaURL string `mapstructure:"schema_url"`
```
I still don't understand the purpose of this.

The input data is going to contain a SchemaURL on its own. What happens if the SchemaURL in the input is different from the one specified here? Are we trying to match this value with the one in the input? Is this for  preloading the schema file? What if the input schema is different than what is specified here?

Instead, why are we not simply not using the SchemaURL in the input data, load that schema file on demand (we can cache it too, even persistently) and do our work using that schema file?

MovieStoreGuy replied:
Wait, I think I may have misunderstood this from the start, I presumed the schema processor had no idea on the location of where the translations files were stored and were part of the configuration so that organisations or vendors could also use this in the event that they require an extension of what exist in the existing semantic convention.

Sorry, I should of laid out my assumptions earlier.

tigrannajaryan commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,24 @@
+Supported Pipelines: traces, metrics, logs
+
+The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration. 
+The latest schema translation file will be bundled as part of the collector build to avoiding having to fetch it at runtime. 
```
The "latest" changes over time, specification publishes new versions and the "latest" schema release in Otel may happen after the particular bundle of the Collector is released. What happens in that case?
I think the Collector should be able to fetch the schema file on demand. Bundled "latest schema available at the release time may be an optimization to speed up Collector startup (but I am not totally sure it is necessary).

MovieStoreGuy replied:
I think it is reasonable to be not in complete lockstep, having the context on your idea of how you expect the processor to work, missing a schema translation file could be resolved at runtime.

However, this idea is a rather round about way and there is currently nothing set up to do the automation.

I should be clear that I read this idea from the spec for it and potentially considering moving this automation til after the ability to transform these signals is complete.

tigrannajaryan commented:
To expand a bit on my earlier comments. Here is how I would expect this processor to work:

- The user specified the "target" schema_url to convert incoming data to.
- Incoming data that has a schema_url specified that belongs to the same family as the "target" will be transformed to that "target". This includes conversion both from older to newer versions and from newer to older versions.
- Incoming data that does not have a schema_url specified or has one that that belongs to a different family is not altered by this processor in any way and passed through.
- To convert from input v1 to target v2 schema where v1<v2 the processor needs schema v2. To convert from v1 to v2 where v1>v2 the processor needs schema v1. Regardless of which schema the processor needs it downloads the schema file on demand, the first time it needs to perform a transformation but does not have the required schema file.
- While download is in progress the processor blocks the processing of that particular batch of pdata in the `Consume*` call. Other concurrent `Consume*` calls may proceed if the schema file they need is already downloaded. 
- Downloaded schema files are cached in memory.
- Optionally: if a `storage` extension is detected download schema files are persisted to the storage. On Collector startup the processor loads previously persisted schema files from the storage.
- Optionally: we can bundle the latest schema file with the Collector. This will speedup the first encounter when input v1 < target v2 (but won't help input v1 > target v2 case).
- Optionally for the future: have filters where the transformation is applied only when the input schema_url matches a particular pattern (this may be useful if for example we know Collector's destination has a problem with a particular schema version, but works fine for other versions).

MovieStoreGuy commented:
Right, so I understand where we are off base. I presumed the processor was unaware of the schema definitions and would allow for the user to use the OpenTelemetry defined schema translation files or potentially use an internally adopt translation translation files (or even vendor's specific files).

Perhaps this may be better written as a `PreCache` field so if these translation can be fetched and collector start instead during the runtime and not blocking data processing (seems that feels like a bad time).

The expected kind of configuration would look like:
```yaml
schema:
  precache:
  - https://opentelemetry.io/schemas/<latest known version at publishing> # Assume this is the default 
  - https://internal.foorbar.org/telemetry/path/to/schemas/<latest>
  
  target_url: # Represents a set of url --> version
  - https://opentelemetry.io/schemas/1.2.0
  - https://internal.foobar.org/telemetry/schemas/1.1.0
```

This feels more like an extension more than anything but this is what I was thinking about writing this.

For the time being, I'll keep the above as food for thought and continue with what you had put down since I also had the same ideas going into it :)

tigrannajaryan commented:
> I presumed the processor was unaware of the schema definitions and would allow for the user to use the OpenTelemetry defined schema translation files or potentially use an internally adopt translation translation files (or even vendor's specific files).

The processor does not need a built-in knowledge about concrete Schema Files. The Otel spec requires that all Schema Files are downloadable at their Schema URL. So the processor should be able to download and use the Schema Files any time it encounters a previously unseen Schema File in the input. This should work for both Otel's Schema Files and for vendor specific Schema Files.

>Perhaps this may be better written as a PreCache field so if these translation can be fetched and collector start instead during the runtime and not blocking data processing (seems that feels like a bad time).

Yes, this is not a bad idea. It may not help a lot since as soon as the Collector is started it may immediately receive some input data and there just may not be enough gap between the startup and input data arrival for the Collector to complete the pre-caching. I guess there is no harm in trying though, as soon as we don't block the Colletor's startup while waiting for pre-caching to complete (the pre-caching needs to proceed asynchronously to avoid blocking the whole startup).

I like how you allow multiple entires in `target_url`. Obviously we must check that only one URL per Schema Family is specified in this section, i.e. you cannot specify both `https://opentelemetry.io/schemas/1.2.0` and `https://opentelemetry.io/schemas/1.5.0` at the same time.

BTW, there is one more scenario: the processor running in an air-gapped environment where it has no ability to download the Schema Files on the fly. In this case your suggestion to bundle the Schema Files can be very useful.

MovieStoreGuy commented:
> The processor does not need a built-in knowledge about concrete Schema Files. The Otel spec requires that all Schema Files are downloadable at their Schema URL. So the processor should be able to download and use the Schema Files any time it encounters a previously unseen Schema File in the input. This should work for both Otel's Schema Files and for vendor specific Schema Files.

Sounds good to me, I skipped over the fact that schema urls are where the translation files are stored as well. 
In the event that there is an issue trying to fetch the configuration, I am going to presume the default path which would be do nothing but log that there was an error for the time being; feels like a safe way to handle progress.

> Yes, this is not a bad idea. It may not help a lot since as soon as the Collector is started it may immediately receive some input data and there just may not be enough gap between the startup and input data arrival for the Collector to complete the pre-caching. I guess there is no harm in trying though, as soon as we don't block the Colletor's startup while waiting for pre-caching to complete (the pre-caching needs to proceed asynchronously to avoid blocking the whole startup).

Yeah, that sounds good to me. Apologies again for not making it clear as to what I was thinking.

> I like how you allow multiple entires in target_url. Obviously we must check that only one URL per Schema Family is specified in this section, i.e. you cannot specify both https://opentelemetry.io/schemas/1.2.0 and https://opentelemetry.io/schemas/1.5.0 at the same time.

On that we can potentially do the following to help enforce on a configuration level that it can't be more than one entry:
```yaml
schema:
  # Additional configuration here 
  targets:
     https://opentelemetry.io/schemas : "1.9.0"
     https://special.vendor.io/telemetry: "1.4.0"
```

> BTW, there is one more scenario: the processor running in an air-gapped environment where it has no ability to download the Schema Files on the fly. In this case your suggestion to bundle the Schema Files can be very useful.

Yeah, my thoughts on this would be to use something like embed so it can be added into the binary that is published by open telemetry. Obviously there is a lot of processes that would need to happen in order to make it work and it could be something similar to the internal published semconv version.

tigrannajaryan commented:
> Sounds good to me, I skipped over the fact that schema urls are where the translation files are stored as well.
> In the event that there is an issue trying to fetch the configuration, I am going to presume the default path which would be do nothing but log that there was an error for the time being; feels like a safe way to handle progress.

Yes, sounds good to me.

MovieStoreGuy commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@
 
 - `hostmetricsreceiver`: Use cpu times for time delta in cpu.utilization calculation (#8856)
 - `dynatraceexporter`: Remove overly verbose stacktrace from certain logs (#8989)
+- `prometheusreceiver`: Fix issues with relabelling the `job` and `instance` labels. (#8780)
```
Moved this in the correct spot

tigrannajaryan commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,35 @@
+
+Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.
+In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;
+using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schames/1.8.2` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.2` so that all existing visualisations will remain unchanged.
```
```suggestion
using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schames/1.8.2` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.2`.
```
I think mentioning the reason why you would want to do a translation is nice to specify, but perhaps a better place for it would be the top of this document, after line 7.

MovieStoreGuy replied:
That makes sense to me!

tigrannajaryan commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,35 @@
+In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;
+using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schames/1.8.2` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.2` so that all existing visualisations will remain unchanged.
+
+Furthemore, it is also possible for organisations and vendors to publish their own semantic convention and be used by this processor, be sure to follow [schema overview](https://opentelemetry.io/docs/reference/specification/schemas/overview/) for all the details.
```
I would also move this line somewhere closer to the top of the file.

tigrannajaryan commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,35 @@
+        precache:
+        - https://opentelemetry.io/schemas/1.9.0
+        targets:
+        - https://opentelemetry.io/schemas/1.6.2
```
Nit: let's use URLs that actually exist:
```suggestion
        - https://opentelemetry.io/schemas/1.6.1
```

MovieStoreGuy replied:
Question for you, is there an easier way for a developer / curious person to see the list of published schemas?

tigrannajaryan replied:
Yes, all Otel schema file originals are here: https://github.com/open-telemetry/opentelemetry-specification/tree/main/schemas

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,98 @@
+		if err != nil {
+			return err
+		}
+		if _, exist := hosts[u.Hostname()]; exist {
```
We shouldn't be using the host name only. We should use the family, which [is defined as](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/overview.md#schema-url):
>The part of the URL preceding the version is called Schema Family identifier. All schemas in one Schema Family have identical Schema Family identifiers.

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+// Config defines the user provided values for the Schema Processor
+type Config struct {
```
Config struct LGTM.

tigrannajaryan reviewed (commented):
Nice! I think we are now moving in the right direction.

tigrannajaryan commented:
Let me know when this is ready for another round of review.

I think we can merge this PR without implementing the actual transformation code to the PR size small and follow up by another PR that adds that transformation logic.

MovieStoreGuy commented:
Hey @tigrannajaryan,

Happy with the initial commit of this, rebased a few times to help clear out some of the conflicts and tying to fix said conflicts.

tigrannajaryan commented on internal/components/components.go:
```diff
@@ -299,6 +300,7 @@ func Components() (component.Factories, error) {
 		spanprocessor.NewFactory(),
 		cumulativetodeltaprocessor.NewFactory(),
 		deltatorateprocessor.NewFactory(),
+		schemaprocessor.NewFactory(),
```
We typically enable the component after the functionality is working. Let's remove the line for now, can be added after the transform code is written.

tigrannajaryan commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+Supported Pipelines: traces, metrics, logs
+
+The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration.
+The processors works be using a set of target schema URLs that are used to match incoming signals. 
```
```suggestion
The processor works by using a set of target schema URLs that are used to match incoming signals. 
```

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+
+	// PreCache is a list of schema URLs that are downloaded
+	// and cached at the start of the collector runtime
+	// in order to avoid fetching data that later on that could
+	// blocking processing of signals. (Optional field)
```
```suggestion
	// in order to avoid fetching data that later on could
	// block processing of signals. (Optional field)
```

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+	// blocking processing of signals. (Optional field)
+	PreCache []string `mapstructure:"precache"`
+
+	// Targets define what schema familys should be
```
```suggestion
	// Targets define what schema families should be
```

tigrannajaryan commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+		return fmt.Errorf("no schema targets defined: %w", errRequiresTargets)
+	}
+
+	familys := make(map[string]struct{})
```
Nit:
```suggestion
	families := make(map[string]struct{})
```

tigrannajaryan commented on processor/schemaprocessor/internal/schema/indentifier.go:
```diff
@@ -0,0 +1,112 @@
+	}, nil
+}
+
+func (id *Identifier) LessOrEqualThan(b *Identifier) bool {
```
1. Is this used anywhere?
2. If I am reading the code correctly the logic does not seem correct. Will it return true for "1.5.0<=1.2.0"?

tigrannajaryan commented:
@open-telemetry/collector-contrib-approvers I am willing to be the sponsor for this component but I am no longer an approver in this repo. If that doesn't work anyone else wants to be a sponsor?

tigrannajaryan reviewed (commented):
Mostly LGTM, just minor nits remaining.

tigrannajaryan commented:
>This branch has conflicts that must be resolved

tigrannajaryan reviewed (approved):
LGTM, but needs a rebase.

MovieStoreGuy commented:
Looks like there is a flake inside a different component :sigh: 
Can't win

tigrannajaryan commented:
We need to wait for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9257 to be merged and rebase so that the test no longer fails.

tigrannajaryan commented:
Please rebase, the failing test is disabled now.

jpkrohling commented on CHANGELOG.md:
```diff
@@ -58,6 +58,8 @@
 
 ### 🚀 New components 🚀
 
+- `schemaprocessor`: Starting the initial work to allow from translating from semantic convention to another (#8371)
```
I would rebase this if I could, but we had another release and this needs to be moved up again. Sorry about this.

jpkrohling commented on cmd/configschema/go.mod:
```diff
@@ -316,7 +316,8 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor v0.48.0 // indirect
-	github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor v0.0.0-00010101000000-000000000000 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor v0.0.0-00010101000000-000000000000 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor v0.48.0 // indirect
```
Same here.

jpkrohling commented on go.mod:
```diff
@@ -75,7 +75,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourceprocessor v0.48.0
 	github.com/open-telemetry/opentelemetry-collector-contrib/processor/routingprocessor v0.48.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor v0.0.0-00010101000000-000000000000
+	github.com/open-telemetry/opentelemetry-collector-contrib/processor/spanmetricsprocessor v0.48.0
```
And here.

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Caching Schema Translation Files
+
+In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing
```
Is it too late to rename `precache` to `prefetch`? I think it's more widely used for similar features elsewhere.

MovieStoreGuy replied:
Makes sense to me :D

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.
+In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;
+using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.
```
```suggestion
using the configuration below, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.
```

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Targets Schemas
+
+Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.
```
```suggestion
Targets define a list of schema URLs with a schema identifier that will be used to translate any schema URL that matches the target URL to that version.
```

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+## Targets Schemas
+
+Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.
+In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;
```
```suggestion
In the event that the processor matches a signal to a target, the processor will translate the signal from the published one to the defined identifier;
```

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+## Caching Schema Translation Files
+
+In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing
+the translations needed for signals that match the schema url.
```
```suggestion
the translations needed for signals that match the schema URL.
```

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration.
+The processor works by using a set of target schema URLs that are used to match incoming signal.
+On a match, the processor will fetch the schema translation file (if not cached) set by the incoming signal and apply transformations
```
```suggestion
On a match, the processor will fetch the schema translation file (if not cached) set by the incoming signal and apply the transformations
```

jpkrohling commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+	// Targets define what schema families should be
+	// translated to, allowing older and newer formats
+	// to conform to the target schema indentifier.
+	Targets []string `mapstructure:"targets"`
```
What happens when I have two targets with conflicting conventions? Is the first taking precedence? The last? The latest following semvem?

MovieStoreGuy replied:
An error is returned during the validation step if there is duplicate schema families to avoid this.

jpkrohling replied:
Could you document this as part of the readme, possibly with examples?

jpkrohling commented on processor/schemaprocessor/factory.go:
```diff
@@ -0,0 +1,107 @@
+
+// NewDefaultConfiguration returns the configuration for schema transformer processor
+// with the default values being used throughout it
+func NewDefaultConfiguration() config.Processor {
```
Any reason this is a public function?

MovieStoreGuy replied:
I was thinking of leaving it public to call inside the processor tests but I think I can get around it using the factory to return the default configuration.

jpkrohling commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+
+func (c *Config) Validate() error {
+	for _, schemaURL := range c.PreCache {
+		_, _, err := schema.GetFamilyAndIdentifier(schemaURL)
```
This is not making an HTTP call at this point, right?

MovieStoreGuy replied:
No, these are validating that the SchemaURL format matches what is expected. 

Actually fetching the data from the internet will be done at `component.Start`

jpkrohling commented on processor/schemaprocessor/testdata/config.yml:
```diff
@@ -0,0 +1,36 @@
+    # to the defined schema identifier set in the target (ie 1.4.2 for all opentelemetry.io/schemas/)
+    targets:
+    - https://opentelemetry.io/schemas/1.4.2
+    - https://internal.fobar.org/otel/schemas/1.2.0
```
```suggestion
    - https://example.com/otel/schemas/1.2.0
```

Let's follow RFC 2606 :-)

MovieStoreGuy committed:
Adding new component: Schema Processor

This is the initial commit into the collector to add in the component to
help enforce a standard semantic convention for all signals.

codeboten commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+Targets define a list of schema URLs with a schema identifier that will be used to translate any schema URL that matches the target URL to that version.
+In the event that the processor matches a signal to a target, the processor will translate the signal from the published one to the defined identifier;
+using the configuration below, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.
```
Wouldn't the example below translate signals published with *any* schema to 1.6.1? I'm not seeing where 1.8.0 is specified anywhere.

MovieStoreGuy replied:
I should have added "for example"

codeboten commented on processor/schemaprocessor/factory.go:
```diff
@@ -0,0 +1,107 @@
+
+var processorCapabilities = consumer.Capabilities{MutatesData: true}
+
+// Factory will store any of the precompiled schemas in future
```
```suggestion
// factory will store any of the precompiled schemas in future
```

MovieStoreGuy replied:
I am honestly surprised that this wasn't caught by the linter, was that intentional?

codeboten commented on processor/schemaprocessor/testdata/config.yml:
```diff
@@ -0,0 +1,36 @@
+    # Targets is a required field that will enable
+    # the processor to convert all telemetry sent 
+    # via the semantic convention family (ie. opentelemetry.io/schemas/*)
+    # to the defined schema identifier set in the target (ie 1.4.2 for all opentelemetry.io/schemas/)
```
Please add a comment about the example.com family as well

codeboten reviewed (commented):
Just a couple of suggestions and questions. Please add a note about what constitutes a family in the readme file, it can be a copy paste of the definition in the schema overview, but i think it will help users of this processor

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+On a match, the processor will fetch the schema translation file (if not cached) set by the incoming signal and apply the transformations
+required to export as the target semantic convention version.
+
+Furthemore, it is also possible for organisations and vendors to publish their own semantic convention and be used by this processor, 
```
```suggestion
Furthermore, it is also possible for organisations and vendors to publish their own semantic conventions and be used by this processor, 
```

jpkrohling commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+Supported Pipelines: traces, metrics, logs
+
+The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration.
+The processor works by using a set of target schema URLs that are used to match incoming signal.
```
```suggestion
The processor works by using a set of target schema URLs that are used to match the incoming signals.
```

jpkrohling reviewed (dismissed):
LGTM. Just clarify to users what happens when conflicting versions of the schemas are used, and fix a couple of typos.

jpkrohling commented:
Sorry, I'm retracting my approval because of the unwanted changes to the changelog.

MovieStoreGuy commented on CHANGELOG.md:
```diff
@@ -63,25 +64,6 @@ https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9278.
-- `windowsperfcountersreceiver`: Move code into separate package for use in other windowsperfcounter receivers (#9108)
-- `datadogexporter`: Add `host_metadata` configuration section to configure host metadata export (#9100)
-- `cmd/mdatagen`: Update documentation generated for attributes to list enumerated values and show the "value" that will be visible on metrics when it is different from the attribute key in metadata.yaml (#8983)
-- `routingprocessor`: add option to drop resource attribute used for routing (#8990)
```
Too easy, I have gotten in habit of checking out main's changelog and adding where my part needs to go.

MovieStoreGuy commented on processor/schemaprocessor/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Caching Schema Translation Files
+
+In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing
```
Not sure if that was a mistake, but `misspell` is reporting it as a typo.

jpkrohling commented on processor/schemaprocessor/config.go:
```diff
@@ -0,0 +1,74 @@
+
+	// Targets define what schema families should be
+	// translated to, allowing older and newer formats
+	// to conform to the target schema indentifier.
```
```suggestion
	// to conform to the target schema identifier.
```

jpkrohling commented on processor/schemaprocessor/internal/schema/identifier_test.go:
```diff
@@ -0,0 +1,89 @@
+			ident:    nil,
+		},
+		{
+			scenario: "schema indentifier with non numerical value",
```
```suggestion
			scenario: "schema identifier with non numerical value",
```

jpkrohling commented on processor/schemaprocessor/transformer_test.go:
```diff
@@ -0,0 +1,110 @@
+	assert.NoError(t, trans.start(context.Background(), nil))
+}
+
+func TestTrasnformerProcessing(t *testing.T) {
```
```suggestion
func TestTransformerProcessing(t *testing.T) {
```

jpkrohling commented on processor/schemaprocessor/transformer_test.go:
```diff
@@ -0,0 +1,110 @@
+	t.Run("logs", func(t *testing.T) {
+		in := pdata.NewLogs()
+		in.ResourceLogs().AppendEmpty()
+		in.ResourceLogs().At(0).SetSchemaUrl("http://opentelemetry.io/schames/1.9.0")
```
```suggestion
		in.ResourceLogs().At(0).SetSchemaUrl("http://opentelemetry.io/schemas/1.9.0")
```

jpkrohling commented on processor/schemaprocessor/transformer_test.go:
```diff
@@ -0,0 +1,110 @@
+	t.Run("traces", func(t *testing.T) {
+		in := pdata.NewTraces()
+		in.ResourceSpans().AppendEmpty()
+		in.ResourceSpans().At(0).SetSchemaUrl("http://opentelemetry.io/schames/1.9.0")
```
```suggestion
		in.ResourceSpans().At(0).SetSchemaUrl("http://opentelemetry.io/schemas/1.9.0")
```

jpkrohling commented on processor/schemaprocessor/transformer_test.go:
```diff
@@ -0,0 +1,110 @@
+	t.Run("metrics", func(t *testing.T) {
+		in := pdata.NewMetrics()
+		in.ResourceMetrics().AppendEmpty()
+		in.ResourceMetrics().At(0).SetSchemaUrl("http://opentelemetry.io/schames/1.9.0")
```
```suggestion
		in.ResourceMetrics().At(0).SetSchemaUrl("http://opentelemetry.io/schemas/1.9.0")
```

jpkrohling reviewed (commented):
Apart from the couple of typos, LGTM!

MovieStoreGuy commented:
@jpkrohling 🙏🏽 

Could I get this merged please?

jpkrohling commented:
I tried to update this branch, but got an error :-( The changelog needs updating, and there's a conflict with the go.sum versions. Once this is fixed and the build is green, I'll merge (if this happens today).

codeboten reviewed (approved):
Thanks for addressing my comments, please rebase and we'll get this merge right away

MovieStoreGuy committed:
Update CHANGELOG.md

MovieStoreGuy committed:
Update CHANGELOG.md

tigrannajaryan commented:
Thank you @MovieStoreGuy 
Looking forward to the continuation PRs!

tigrannajaryan commented:
@MovieStoreGuy do you plan to continue working on the schema processor?

MovieStoreGuy commented:
Yes, sorry. I have been meaning to get back into it but I have been working on rollout infra changes to adopt the otel collector for resource metrics, so that has taken a lot of brain space atm.

I intend to get back to this this week
